File: android/app/src/main/java/com/podswitch/platform/PresenceCoordinator.kt
package com.podswitch.platform

import android.content.Context
import android.net.ConnectivityManager
import android.net.wifi.WifiManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.podswitch.core.ClaimRegistry
import com.podswitch.core.PresenceMessage
import com.podswitch.core.PresencePort
import java.net.DatagramPacket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.MulticastSocket
import java.net.NetworkInterface
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

/**
 * UDP-multicast LAN coordination. Broadcasts a CLAIM while this device holds + plays the target,
 * listens for peer claims, and exposes whether a peer is currently active. Best-effort: if the
 * network blocks multicast or no peer is present, [peerActiveOnTarget] stays false and the engine
 * falls back to the reactive yield guard.
 *
 * Auto-paired: messages are authenticated with an HMAC keyed on the normalized target address, so
 * two devices targeting the same earbuds coordinate with no typed setup, and nobody else can.
 */
class AndroidPresenceCoordinator(
    context: Context,
    private val deviceId: String,
    /** Returns the raw configured target Bluetooth address, or null if unconfigured. */
    private val targetProvider: () -> String?,
) : PresencePort {

    override var onPeerChanged: (() -> Unit)? = null

    private val appContext = context.applicationContext
    private val wifi = appContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
    private val connectivity =
        appContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    private val mainHandler = Handler(Looper.getMainLooper())
    private val registry = ClaimRegistry(deviceId)

    private val running = AtomicBoolean(false)
    private var socket: MulticastSocket? = null
    private var multicastLock: WifiManager.MulticastLock? = null
    private var receiverThread: Thread? = null
    /** Subnet-directed broadcast address for the Wi-Fi interface, used as a fallback delivery path
     *  alongside multicast. Null if no Wi-Fi interface/broadcast address could be resolved. */
    private var broadcastAddress: InetAddress? = null

    private val sender = Executors.newSingleThreadScheduledExecutor()
    private var heartbeat: java.util.concurrent.ScheduledFuture<*>? = null
    private var releaseTimer: java.util.concurrent.ScheduledFuture<*>? = null
    private var localActive = false

    private var lastPeerActive = false

    /** Open the socket, join the group, and start listening. Safe to call once. */
    fun start() {
        if (!running.compareAndSet(false, true)) return
        try {
            multicastLock = wifi.createMulticastLock("podswitch").apply {
                setReferenceCounted(false)
                acquire()
            }
            val group = InetAddress.getByName(GROUP)
            val wifiNif = resolveWifiInterface()
            socket = MulticastSocket(PORT).apply {
                reuseAddress = true
                broadcast = true
                if (wifiNif != null) {
                    networkInterface = wifiNif
                    joinGroup(InetSocketAddress(group, PORT), wifiNif)
                    Log.i(TAG, "socket opened, joined $GROUP:$PORT on interface ${wifiNif.name}")
                } else {
                    @Suppress("DEPRECATION")
                    joinGroup(group)
                    Log.w(TAG, "no Wi-Fi NetworkInterface resolved; joined $GROUP:$PORT unbound")
                }
                timeToLive = 1
            }
            broadcastAddress = resolveBroadcastAddress(wifiNif)
            receiverThread = Thread({ receiveLoop() }, "podswitch-presence-rx").also { it.start() }
        } catch (t: Throwable) {
            Log.e(TAG, "start() failed", t)
            stop()
        }
    }

    /** Stop heartbeating, leave the group, and release resources. */
    fun stop() {
        running.set(false)
        cancelHeartbeat()
        cancelReleaseTimer()
        runCatching { socket?.close() }
        socket = null
        runCatching { multicastLock?.release() }
        multicastLock = null
        receiverThread = null
    }

    fun shutdown() {
        stop()
        sender.shutdownNow()
    }

    override fun peerActiveOnTarget(): Boolean =
        synchronized(registry) { registry.peerActive(System.currentTimeMillis()) }

    override fun setLocalActive(active: Boolean) {
        sender.execute {
            if (active == localActive) return@execute
            localActive = active
            if (active) {
                Log.d(TAG, "local active=true, sending CLAIM + starting heartbeat")
                cancelReleaseTimer()
                sendNow(PresenceMessage.Type.CLAIM)
                heartbeat = sender.scheduleAtFixedRate(
                    { sendNow(PresenceMessage.Type.CLAIM) },
                    HEARTBEAT_MS, HEARTBEAT_MS, java.util.concurrent.TimeUnit.MILLISECONDS,
                )
            } else {
                // Don't send RELEASE immediately: a brief pause (e.g. between tracks) would hand
                // the target to a peer that's only reacting to a momentary gap. Instead, stop
                // heartbeating now and send RELEASE after a debounce -- long enough to ride out a
                // track change, but still short enough to proactively notify the peer of a genuine
                // stop (rather than relying solely on the passive TTL, which never fires if no
                // further packets arrive to trigger a re-check).
                Log.d(TAG, "local active=false, stopping heartbeat, scheduling delayed RELEASE")
                cancelHeartbeat()
                releaseTimer = sender.schedule(
                    {
                        Log.d(TAG, "debounce elapsed, sending RELEASE")
                        sendNow(PresenceMessage.Type.RELEASE)
                        releaseTimer = null
                    },
                    RELEASE_DEBOUNCE_MS, java.util.concurrent.TimeUnit.MILLISECONDS,
                )
            }
        }
    }

    private fun cancelHeartbeat() {
        heartbeat?.cancel(false)
        heartbeat = null
    }

    private fun cancelReleaseTimer() {
        releaseTimer?.cancel(false)
        releaseTimer = null
    }

    private fun sendNow(type: PresenceMessage.Type) {
        val target = targetProvider()?.let { normalize(it) } ?: return
        val sock = socket ?: return
        val message = PresenceMessage(
            type = type,
            deviceId = deviceId,
            target = target,
            playing = type == PresenceMessage.Type.CLAIM,
            timestamp = System.currentTimeMillis(),
            ttlMillis = TTL_MS,
        )
        val bytes = message.encode()
        runCatching {
            sock.send(DatagramPacket(bytes, bytes.size, InetAddress.getByName(GROUP), PORT))
        }.onFailure { Log.w(TAG, "send to multicast group failed", it) }

        val bcast = broadcastAddress
        if (bcast != null) {
            runCatching {
                sock.send(DatagramPacket(bytes, bytes.size, bcast, PORT))
            }.onFailure { Log.w(TAG, "send to subnet broadcast failed", it) }
        }
        runCatching {
            sock.send(DatagramPacket(bytes, bytes.size, InetAddress.getByName("255.255.255.255"), PORT))
        }.onFailure { Log.w(TAG, "send to 255.255.255.255 failed", it) }
    }

    private fun receiveLoop() {
        val buffer = ByteArray(2048)
        while (running.get()) {
            val sock = socket ?: break
            val packet = DatagramPacket(buffer, buffer.size)
            val received = runCatching { sock.receive(packet); true }
                .onFailure { if (running.get()) Log.w(TAG, "receive() failed", it) }
                .getOrDefault(false)
            if (!received) break

            // Log the sender before the target/HMAC filter so a two-device test can tell "nothing
            // arrived" (network problem) apart from "arrived but rejected" (decode/logic problem).
            val rawSenderId = runCatching {
                String(packet.data, 0, packet.length, Charsets.UTF_8).split("|").getOrNull(2)
            }.getOrNull()
            Log.d(TAG, "received datagram (${packet.length} bytes) from deviceId=$rawSenderId")

            val target = targetProvider()?.let { normalize(it) }
            if (target == null) {
                Log.d(TAG, "dropping datagram: no target configured")
                continue
            }
            val message = PresenceMessage.decodeAndVerify(packet.data, packet.length, target)
            if (message == null) {
                Log.d(TAG, "dropping datagram: decode/HMAC verification failed")
                continue
            }
            Log.d(TAG, "decoded ${message.type} from ${message.deviceId} playing=${message.playing}")
            val now = synchronized(registry) {
                registry.record(message, System.currentTimeMillis())
                registry.peerActive(System.currentTimeMillis())
            }
            if (now != lastPeerActive) {
                lastPeerActive = now
                Log.i(TAG, "peerActive -> $now")
                mainHandler.post { onPeerChanged?.invoke() }
            }
        }
    }

    /** Resolve the [NetworkInterface] backing the currently active Wi-Fi network, if any. Pinning
     *  the socket to it avoids multicast silently binding to cellular when both are up. */
    private fun resolveWifiInterface(): NetworkInterface? {
        return runCatching {
            val network = connectivity.activeNetwork ?: return null
            val linkProperties = connectivity.getLinkProperties(network) ?: return null
            val name = linkProperties.interfaceName ?: return null
            NetworkInterface.getByName(name)
        }.onFailure { Log.w(TAG, "resolveWifiInterface failed", it) }.getOrNull()
    }

    private fun resolveBroadcastAddress(nif: NetworkInterface?): InetAddress? {
        val iface = nif ?: return null
        return runCatching {
            iface.interfaceAddresses.firstOrNull { it.broadcast != null }?.broadcast
        }.onFailure { Log.w(TAG, "resolveBroadcastAddress failed", it) }.getOrNull()
    }

    private fun normalize(address: String): String =
        address.lowercase().replace(":", "").replace("-", "")

    private companion object {
        const val TAG = "PodSwitchPresence"
        const val GROUP = "239.7.7.7"
        const val PORT = 54321
        const val HEARTBEAT_MS = 2000L
        const val TTL_MS = 6000L
        const val RELEASE_DEBOUNCE_MS = 4000L
    }
}

File: macos/Sources/PodSwitchCore/PresenceCoordinator.swift
import Foundation
import Darwin
import os

/// UDP-multicast LAN coordination for macOS, interoperating byte-for-byte with the Android app.
///
/// Broadcasts a CLAIM while this Mac holds + plays the target, listens for peer claims, and exposes
/// whether a peer is active. Uses raw POSIX multicast sockets rather than Network.framework, whose
/// multicast API requires an entitlement unavailable to ad-hoc-signed builds. macOS 15+ shows a
/// one-time Local Network permission prompt; if denied, no peers are heard and the engine falls
/// back to the reactive yield guard.
public final class PresenceCoordinator: PresencePort, @unchecked Sendable {

    private let deviceId: String
    private let queue = DispatchQueue(label: "com.podswitch.core.presence")
    private let lock = NSLock()
    private let logger = Logger(subsystem: "com.podswitch.core", category: "Presence")

    private var _onPeerChanged: (@Sendable () -> Void)?
    public var onPeerChanged: (@Sendable () -> Void)? {
        get { lock.lock(); defer { lock.unlock() }; return _onPeerChanged }
        set { lock.lock(); _onPeerChanged = newValue; lock.unlock() }
    }

    // Queue-confined state.
    private let registry: ClaimRegistry
    private var fd: Int32 = -1
    private var readSource: DispatchSourceRead?
    private var heartbeat: DispatchSourceTimer?
    private var releaseTimer: DispatchSourceTimer?
    private var localActive = false
    private var lastPeerActive = false
    private var normalizedTarget: String?
    /// Subnet-directed broadcast address for the pinned interface (e.g. 192.168.1.255), used as a
    /// fallback delivery path alongside multicast. Nil if no suitable interface was found.
    private var broadcastAddress: String?

    public init(deviceId: String) {
        self.deviceId = deviceId
        self.registry = ClaimRegistry(ownDeviceId: deviceId)
    }

    /// Update the configured target (raw BT address). Call from anywhere; confined to the queue.
    public func updateTarget(_ rawAddress: String?) {
        let normalized = rawAddress.map { Self.normalize($0) }
        queue.async { [weak self] in self?.normalizedTarget = normalized }
    }

    /// Open the socket, join the group, and start listening.
    public func start() {
        queue.async { [weak self] in self?.openSocket() }
    }

    /// Stop heartbeating, leave the group, and release the socket.
    public func stop() {
        queue.async { [weak self] in self?.closeSocket() }
    }

    public func peerActiveOnTarget() -> Bool {
        queue.sync { registry.peerActive(nowMillis: Self.nowMillis()) }
    }

    public func setLocalActive(_ active: Bool) {
        queue.async { [weak self] in
            guard let self, active != self.localActive else { return }
            self.localActive = active
            if active {
                self.logger.debug("local active=true, sending CLAIM + starting heartbeat")
                self.releaseTimer?.cancel()
                self.releaseTimer = nil
                self.send(.claim)
                let timer = DispatchSource.makeTimerSource(queue: self.queue)
                timer.schedule(deadline: .now() + Self.heartbeat, repeating: Self.heartbeat)
                timer.setEventHandler { [weak self] in self?.send(.claim) }
                self.heartbeat = timer
                timer.resume()
            } else {
                // Don't send RELEASE immediately: a brief pause (e.g. between tracks) would hand
                // the target to a peer that's only reacting to a momentary gap. Instead, stop
                // heartbeating now and send RELEASE after a debounce -- long enough to ride out a
                // track change, but still short enough to proactively notify the peer of a genuine
                // stop (rather than relying solely on the passive TTL, which never fires if no
                // further packets arrive to trigger a re-check).
                self.logger.debug("local active=false, stopping heartbeat, scheduling delayed RELEASE")
                self.heartbeat?.cancel()
                self.heartbeat = nil
                let timer = DispatchSource.makeTimerSource(queue: self.queue)
                timer.schedule(deadline: .now() + Self.releaseDebounce)
                timer.setEventHandler { [weak self] in
                    self?.logger.debug("debounce elapsed, sending RELEASE")
                    self?.send(.release)
                    self?.releaseTimer = nil
                }
                self.releaseTimer = timer
                timer.resume()
            }
        }
    }

    // MARK: - Socket (queue-confined)

    private func openSocket() {
        guard fd < 0 else { return }
        let s = socket(AF_INET, SOCK_DGRAM, 0)
        guard s >= 0 else {
            logger.error("socket() failed errno=\(errno) (\(String(cString: strerror(errno))))")
            return
        }
        var yes: Int32 = 1
        setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &yes, socklen_t(MemoryLayout<Int32>.size))
        setsockopt(s, SOL_SOCKET, SO_REUSEPORT, &yes, socklen_t(MemoryLayout<Int32>.size))
        if setsockopt(s, SOL_SOCKET, SO_BROADCAST, &yes, socklen_t(MemoryLayout<Int32>.size)) != 0 {
            logger.warning("SO_BROADCAST failed errno=\(errno)")
        }

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = in_port_t(UInt16(Self.port).bigEndian)
        addr.sin_addr.s_addr = INADDR_ANY
        let bound = withUnsafePointer(to: &addr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                bind(s, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        guard bound == 0 else {
            logger.error("bind() failed errno=\(errno) (\(String(cString: strerror(errno))))")
            close(s)
            return
        }

        let iface = Self.activeInterface()
        if let iface {
            logger.info("pinning multicast to interface \(iface.name, privacy: .public) addr=\(iface.address, privacy: .private)")
            broadcastAddress = iface.broadcast
        } else {
            logger.warning("no active non-loopback IPv4 interface found; joining on INADDR_ANY")
            broadcastAddress = nil
        }

        var mreq = ip_mreq()
        mreq.imr_multiaddr.s_addr = inet_addr(Self.group)
        mreq.imr_interface.s_addr = iface.map { inet_addr($0.address) } ?? INADDR_ANY
        if setsockopt(s, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, socklen_t(MemoryLayout<ip_mreq>.size)) != 0 {
            logger.error("IP_ADD_MEMBERSHIP failed errno=\(errno) (\(String(cString: strerror(errno))))")
        }

        // IP_ADD_MEMBERSHIP only pins the receive/join side; pin outgoing multicast separately.
        if let iface {
            var ifAddr = in_addr(s_addr: inet_addr(iface.address))
            if setsockopt(s, IPPROTO_IP, IP_MULTICAST_IF, &ifAddr, socklen_t(MemoryLayout<in_addr>.size)) != 0 {
                logger.warning("IP_MULTICAST_IF failed errno=\(errno)")
            }
        }
        var ttl: UInt8 = 1
        setsockopt(s, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, socklen_t(MemoryLayout<UInt8>.size))
        var loop: UInt8 = 1
        setsockopt(s, IPPROTO_IP, IP_MULTICAST_LOOP, &loop, socklen_t(MemoryLayout<UInt8>.size))

        _ = fcntl(s, F_SETFL, O_NONBLOCK)
        fd = s
        logger.info("socket opened fd=\(s) group=\(Self.group, privacy: .public) port=\(Self.port)")

        let source = DispatchSource.makeReadSource(fileDescriptor: s, queue: queue)
        source.setEventHandler { [weak self] in self?.drain() }
        readSource = source
        source.resume()
    }

    private func closeSocket() {
        heartbeat?.cancel(); heartbeat = nil
        releaseTimer?.cancel(); releaseTimer = nil
        readSource?.cancel(); readSource = nil
        if fd >= 0 { close(fd); fd = -1 }
        localActive = false
    }

    private func drain() {
        guard fd >= 0 else { return }
        var buffer = [UInt8](repeating: 0, count: 2048)
        while true {
            let n = recv(fd, &buffer, buffer.count, 0)
            if n < 0 {
                if errno != EWOULDBLOCK && errno != EAGAIN {
                    logger.error("recv() failed errno=\(errno) (\(String(cString: strerror(errno))))")
                }
                break
            }
            if n == 0 { break }
            handle(Data(buffer[0..<n]))
        }
    }

    private func handle(_ data: Data) {
        // Log the sender before the target/HMAC filter so a two-device test can tell "nothing
        // arrived" (network problem) apart from "arrived but rejected" (decode/logic problem).
        let rawSenderId = String(data: data, encoding: .utf8)?
            .components(separatedBy: "|")
            .dropFirst(2).first
        logger.debug("received datagram (\(data.count) bytes) from deviceId=\(rawSenderId ?? "?", privacy: .public)")

        guard let target = normalizedTarget else {
            logger.debug("dropping datagram: no target configured")
            return
        }
        guard let message = PresenceMessage.decodeAndVerify(data, normalizedTarget: target) else {
            logger.debug("dropping datagram: decode/HMAC verification failed")
            return
        }
        logger.debug("decoded \(message.kind.rawValue, privacy: .public) from \(message.deviceId, privacy: .public) playing=\(message.playing)")
        registry.record(message, nowMillis: Self.nowMillis())
        let active = registry.peerActive(nowMillis: Self.nowMillis())
        if active != lastPeerActive {
            lastPeerActive = active
            logger.info("peerActive -> \(active)")
            onPeerChanged?()
        }
    }

    private func send(_ kind: PresenceMessage.Kind) {
        guard fd >= 0, let target = normalizedTarget else { return }
        let message = PresenceMessage(
            kind: kind,
            deviceId: deviceId,
            target: target,
            playing: kind == .claim,
            timestamp: Self.nowMillis(),
            ttlMillis: Self.ttl
        )
        let data = message.encode()
        sendTo(data, address: Self.group)
        if let broadcastAddress {
            sendTo(data, address: broadcastAddress)
        }
        sendTo(data, address: "255.255.255.255")
    }

    private func sendTo(_ data: Data, address: String) {
        var dest = sockaddr_in()
        dest.sin_family = sa_family_t(AF_INET)
        dest.sin_port = in_port_t(UInt16(Self.port).bigEndian)
        dest.sin_addr.s_addr = inet_addr(address)
        let sent = data.withUnsafeBytes { raw in
            withUnsafePointer(to: &dest) {
                $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                    sendto(fd, raw.baseAddress, data.count, 0, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
        }
        if sent < 0 {
            logger.error("sendto \(address, privacy: .public) failed errno=\(errno) (\(String(cString: strerror(errno))))")
        }
    }

    // MARK: - Interface discovery

    private struct Interface {
        let name: String
        let address: String
        let broadcast: String?
    }

    /// The first up, non-loopback, IPv4 interface (typically Wi-Fi `en0`), used to pin outgoing
    /// multicast/broadcast rather than leaving it to `INADDR_ANY`, which can silently bind to the
    /// wrong interface (or none) and drop all peer traffic.
    private static func activeInterface() -> Interface? {
        var ifaddrPtr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddrPtr) == 0, let first = ifaddrPtr else { return nil }
        defer { freeifaddrs(first) }

        var cursor: UnsafeMutablePointer<ifaddrs>? = first
        var candidate: Interface?
        while let entry = cursor {
            defer { cursor = entry.pointee.ifa_next }
            let flags = Int32(entry.pointee.ifa_flags)
            guard flags & IFF_UP != 0, flags & IFF_LOOPBACK == 0,
                  let addr = entry.pointee.ifa_addr, addr.pointee.sa_family == sa_family_t(AF_INET)
            else { continue }

            let name = String(cString: entry.pointee.ifa_name)
            let address = addr.withMemoryRebound(to: sockaddr_in.self, capacity: 1) {
                String(cString: inet_ntoa($0.pointee.sin_addr))
            }
            var broadcast: String?
            if flags & IFF_BROADCAST != 0, let bcastAddr = entry.pointee.ifa_dstaddr,
               bcastAddr.pointee.sa_family == sa_family_t(AF_INET) {
                broadcast = bcastAddr.withMemoryRebound(to: sockaddr_in.self, capacity: 1) {
                    String(cString: inet_ntoa($0.pointee.sin_addr))
                }
            }
            let found = Interface(name: name, address: address, broadcast: broadcast)
            // Prefer en0 (typical Wi-Fi on Mac laptops/desktops) but accept the first candidate
            // found so we still work on machines where Wi-Fi isn't en0.
            if name == "en0" { return found }
            if candidate == nil { candidate = found }
        }
        return candidate
    }

    // MARK: - Helpers

    private static func normalize(_ address: String) -> String {
        address.lowercased().replacingOccurrences(of: ":", with: "").replacingOccurrences(of: "-", with: "")
    }

    private static func nowMillis() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }

    private static let group = "239.7.7.7"
    private static let port = 54321
    private static let heartbeat: DispatchTimeInterval = .seconds(2)
    private static let ttl: Int64 = 6000
    private static let releaseDebounce: DispatchTimeInterval = .seconds(4)
}
